apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -49,3 +55,254 @@
 - Missing $this
 
 */
+
+use std::fmt::Display;
+
+use apollo_compiler::schema::Component;
+use apollo_compiler::schema::Directive;
+use apollo_compiler::Schema;
+use itertools::Itertools;
+use url::Url;
+
+use crate::sources::connect::spec::schema::CONNECT_HTTP_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::CONNECT_SOURCE_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::SOURCE_BASE_URL_ARGUMENT_NAME;
+use crate::sources::connect::spec::schema::SOURCE_NAME_ARGUMENT_NAME;
+
+/// Validate the connectors-related directives `@source` and `@connect`.
+///
+/// This function attempts to collect as many validation errors as possible, so it does not bail
+/// out as soon as it encounters one.
+pub fn validate(schema: Schema) -> Vec<ValidationError> {
+    let mut source_names = Vec::new();
+    let errors = schema
+        .schema_definition
+        .directives
+        .iter()
+        .filter_map(|directive| {
+            if directive.name == CONNECT_SOURCE_ARGUMENT_NAME {
+                match validate_source(directive) {
+                    Ok(source_directive) => {
+                        if let SourceName::Valid(name) = source_directive.name {
+                            source_names.push(name);
+                        }
+                        (!source_directive.errors.is_empty()).then_some(source_directive.errors)
+                    }
+                    Err(error) => Some(vec![error]),
+                }
+            } else {
+                None
+            }
+        })
+        .flatten()
+        .collect_vec();
+
+    if source_names.is_empty() && errors.is_empty() {
+        todo!("Check for renamed directives—this code is _only_ called for connectors subgraphs")
```

Comment:
i think we're going to allow using `@connect(http: { GET: "https://whatever" })` without a source! (i'm hoping it makes the initial demo/evaluation a touch easier.) so an empty list of sources is expected.

Reply:
Oh, right. Okay I moved the TODO into the big block of TODOs instead 😅